File: ada/ada2012/Go/AdaLexerBase.go
package parser

import (
	"github.com/antlr4-go/antlr/v4"
)

type AdaLexerBase struct {
	*antlr.BaseLexer
	_lastTokenType int
}

func (l *AdaLexerBase) NextToken() antlr.Token {
	token := l.BaseLexer.NextToken()
	if token.GetChannel() == antlr.TokenDefaultChannel {
		l._lastTokenType = token.GetTokenType()
	}
	return token
}

func (l *AdaLexerBase) IsCharLiteralAllowed() bool {
	// In Ada, a tick after an identifier, closing paren, or 'all' keyword
	// is an attribute tick, not the start of a character literal.
	return l._lastTokenType != AdaLexerIDENTIFIER_ &&
		l._lastTokenType != AdaLexerRP &&
		l._lastTokenType != AdaLexerALL
}

File: ada/ada2012/Dart/AdaLexerBase.dart
import 'package:antlr4/antlr4.dart';
import 'AdaLexer.dart';

abstract class AdaLexerBase extends Lexer {
    int _lastTokenType = 0;

    AdaLexerBase(CharStream input) : super(input);

    @override
    Token nextToken() {
        var token = super.nextToken();
        if (token.channel == Token.DEFAULT_CHANNEL) {
            _lastTokenType = token.type!;
        }
        return token;
    }

    bool IsCharLiteralAllowed() {
        // In Ada, a tick after an identifier, closing paren, or 'all' keyword
        // is an attribute tick, not the start of a character literal.
        return _lastTokenType != AdaLexer.TOKEN_IDENTIFIER_
            && _lastTokenType != AdaLexer.TOKEN_RP
            && _lastTokenType != AdaLexer.TOKEN_ALL;
    }
}
